File: Demo/HoundifySDK Sample App (Objective-C)/HoundifySDK Sample App (Objective-C)/AudioTester.m
//
//  AudioTester.m
//  HoundSDK Test Application
//
//  Created by Cyril Austin on 6/2/15.
//  Copyright (c) 2015 SoundHound, Inc. All rights reserved.
//

#import "AudioTester.h"

#define OUTPUT_BUS                              0
#define INPUT_BUS                               1

@import AVFoundation;

#pragma mark - Callbacks

typedef void (^AudioTesterPermissionCallback)(BOOL granted);

#pragma mark - Errors

NSString* AudioTesterErrorDomain = @"AudioTesterErrorDomain";

#pragma mark - AudioTester

@interface AudioTester()

@property(nonatomic, strong) dispatch_queue_t queue;
@property(nonatomic, assign) AudioUnit audioUnit;

@property(nonatomic, copy) AudioTesterDataHandler handler;

@end

@implementation AudioTester

+ (instancetype)instance
{
    static AudioTester* instance = nil;
    
    static dispatch_once_t onceToken;
    
    dispatch_once(&onceToken, ^{
    
        instance = [[self alloc] init];
    });
    
    return instance;
}

- (instancetype)init
{
    self = [super init];
    
    if (self)
    {
        self.queue = dispatch_queue_create("com.hound.audio", NULL);
    }
    
    return self;
}

- (void)startAudioWithSampleRate:(double)sampleRate dataHandler:(AudioTesterDataHandler)handler
{
    [self stopAudioWithHandler:^(NSError* error) {
    
        if (error)
        {
            if (handler) handler(error, nil);
        }
        else
        {
            dispatch_async(self.queue, ^{
            
                self.handler = handler;
            
                [self requestPermissionsWithCompletionHandler:^(BOOL granted) {
                    
                    dispatch_async(self.queue, ^{
                
                        NSError* error = nil;
                        
                        if (!granted)
                        {
                            error = [self errorWithCode:AudioTesterErrorCodePermissionDenied];
                        }
                        
                        if (!error)
                        {
                            [self.session setCategory:AVAudioSessionCategoryPlayAndRecord
                                withOptions:AVAudioSessionCategoryOptionAllowBluetooth
                                error:&error];
                        }
                        
                        if (error)
                        {
                            if (handler) handler(error, nil);
                        }
                        else
                        {
                            [self.session setMode:AVAudioSessionModeDefault error:&error];

                            if (!error)
                            {
                                [self.session setPreferredSampleRate:sampleRate error:&error];
                            }
                            
                            if (!error)
                            {
                                [self.session setActive:YES error:&error];
                            }
                            
                            if (error)
                            {
                                if (handler) handler(error, nil);
                            }
                            else
                            {
                                OSStatus status = kAudioServicesNoError;
                                
                                AudioComponentDescription audioComponentDescription;

                                audioComponentDescription.componentType = kAudioUnitType_Output;
                                audioComponentDescription.componentSubType = kAudioUnitSubType_RemoteIO;
                                audioComponentDescription.componentManufacturer = kAudioUnitManufacturer_Apple;
                                audioComponentDescription.componentFlags = 0;
                                audioComponentDescription.componentFlagsMask = 0;
                                
                                AudioComponent audioComponent = AudioComponentFindNext(NULL, &audioComponentDescription);
                                
                                if (!audioComponent)
                                {
                                    status = kAudioServicesUnsupportedPropertyError;
                                }
                                else
                                {
                                    status = AudioComponentInstanceNew(audioComponent, &_audioUnit);
                                    
                                    if (status == kAudioServicesNoError)
                                    {
                                        UInt32 yes = YES;
                                    
                                        if (status == kAudioServicesNoError)
                                        {
                                            status = AudioUnitSetProperty(
                                                self.audioUnit,
                                                kAudioOutputUnitProperty_EnableIO,
                                                kAudioUnitScope_Input,
                                                INPUT_BUS,
                                                &yes,
                                                sizeof(yes));
                                        }

                                        AudioStreamBasicDescription streamDescription = self.streamDescription;
                                        
                                        if (status == kAudioServicesNoError)
                                        {
                                            status = AudioUnitSetProperty(
                                                self.audioUnit,
                                                kAudioUnitProperty_StreamFormat,
                                                kAudioUnitScope_Output,
                                                INPUT_BUS,
                                                &streamDescription,
                                                sizeof(streamDescription));
                                        }

                                        if (status == kAudioServicesNoError)
                                        {
                                            status = AudioUnitSetProperty(
                                                self.audioUnit,
                                                kAudioUnitProperty_StreamFormat,
                                                kAudioUnitScope_Input,
                                                OUTPUT_BUS,
                                                &streamDescription,
                                                sizeof(streamDescription));
                                        }
                                        
                                        AURenderCallbackStruct renderCallbackStruct;
                                        renderCallbackStruct.inputProc = audioTesterRenderCallback;
                                        renderCallbackStruct.inputProcRefCon = (__bridge void*)self;
                                        
                                        if (status == kAudioServicesNoError)
                                        {
                                            status = AudioUnitSetProperty(
                                                self.audioUnit,
                                                kAudioUnitProperty_SetRenderCallback,
                                                kAudioUnitScope_Global,
                                                OUTPUT_BUS,
                                                &renderCallbackStruct,
                                                sizeof(renderCallbackStruct));
                                        }
                                        
                                        if (status == kAudioServicesNoError)
                                        {
                                            status = AudioUnitInitialize(self.audioUnit);
                                        }
                                        
                                        if (status == kAudioServicesNoError)
                                        {
                                            status = AudioOutputUnitStart(self.audioUnit);
                                        }
                                    }
                                }
                                
                                NSError* error = [self errorWithOSStatus:status];
                                
                                if (error)
                                {
                                    if (handler) handler(error, nil);
                                }
                            }
                        }
                    });
                }];
            });
        }
    }];
}

- (void)stopAudioWithHandler:(AudioTesterErrorHandler)handler
{
    dispatch_async(self.queue, ^{
        
        NSError* error = nil;
        
        if (self.audioUnit)
        {
            OSStatus status;
            
            status = AudioOutputUnitStop(self.audioUnit);
            
            if (!error)
            {
                error = [self errorWithOSStatus:status];
            }
        
            status = AudioUnitUninitialize(self.audioUnit);
                
            if (!error)
            {
                error = [self errorWithOSStatus:status];
            }
            
            status = AudioComponentInstanceDispose(self.audioUnit);
            
            if (!error)
            {
                error = [self errorWithOSStatus:status];
            }
            
            self.audioUnit = NULL;

            [self.session setActive:NO error:&error];
        }
        
        if (handler) handler(error);
    });
}

#pragma mark - Audio Processing

OSStatus audioTesterRenderCallback(void* context, AudioUnitRenderActionFlags* actionFlags,
    const AudioTimeStamp* timestamp, UInt32 busNumber, UInt32 frameCount,
    AudioBufferList* bufferList)
{
	AudioTester* self = (__bridge AudioTester*)context;
    
    return [self render:actionFlags timestamp:timestamp
        busNumber:busNumber frameCount:frameCount bufferList:bufferList];
}

- (OSStatus)render:(AudioUnitRenderActionFlags*)actionFlags
    timestamp:(const AudioTimeStamp*)timestamp
	busNumber:(UInt32)busNumber
    frameCount:(UInt32)frameCount
    bufferList:(AudioBufferList*)bufferList
{
    OSStatus status = kAudioServicesNoError;
    
    if (!bufferList)
    {
        status = kAudioServicesUnsupportedPropertyError;
    }
    else
    {
        status = AudioUnitRender(self.audioUnit, actionFlags, timestamp,
            INPUT_BUS, frameCount, bufferList);
        
        if (status == kAudioServicesNoError)
        {
            if (bufferList->mNumberBuffers > 0)
            {
                SInt16* buffer = bufferList->mBuffers[0].mData;
                NSUInteger length = sizeof(SInt16) * frameCount;
                
                NSData* data = [NSData dataWithBytes:buffer length:length];
                
                if (self.handler) self.handler(nil, data);
            }
        }
    }

    *actionFlags |= kAudioUnitRenderAction_OutputIsSilence;

    NSError* error = [self errorWithOSStatus:status];

    if (error)
    {
        if (self.handler) self.handler(error, nil);
    }
    
	return status;
}

#pragma mark - Utility

- (AVAudioSession*)session
{
    return AVAudioSession.sharedInstance;
}

- (NSError*)errorWithCode:(AudioTesterErrorCode)errorCode
{
    NSError* error = [NSError errorWithDomain:AudioTesterErrorDomain
        code:errorCode userInfo:nil];
    
    return error;
}

- (NSError*)errorWithOSStatus:(OSStatus)status
{
    NSError* error = nil;

    if (status != kAudioServicesNoError)
    {
        error = [NSError errorWithDomain:NSOSStatusErrorDomain code:status userInfo:nil];
    }
    
    return error;
}

- (AudioStreamBasicDescription)streamDescription
{
    AudioStreamBasicDescription streamDescription;
    
    streamDescription.mFormatID = kAudioFormatLinearPCM;
    streamDescription.mFormatFlags = kAudioFormatFlagIsSignedInteger | kAudioFormatFlagIsPacked;
    streamDescription.mFramesPerPacket = 1;
    streamDescription.mChannelsPerFrame = 1;
    streamDescription.mBitsPerChannel = 16;
    streamDescription.mBytesPerPacket = 2;
    streamDescription.mBytesPerFrame = 2;
    streamDescription.mSampleRate = [[AVAudioSession sharedInstance] sampleRate];
    
    return streamDescription;
}

- (void)requestPermissionsWithCompletionHandler:(AudioTesterPermissionCallback)handler
{
    AVAudioSessionRecordPermission permission = AVAudioSessionRecordPermissionUndetermined;
    
    if ([self.session respondsToSelector:@selector(recordPermission)])
    {
        permission = [self.session recordPermission];
    }
    
    if (permission == AVAudioSessionRecordPermissionDenied)
    {
        if (handler) handler(NO);
    }
    else if (permission == AVAudioSessionRecordPermissionGranted)
    {
        if (handler) handler(YES);
    }
    else
    {
        [self.session requestRecordPermission:handler];
    }
}

@end

File: Demo/HoundifySDK Sample App (Swift)/HoundifySDK Sample App (Swift)/AudioTester.swift
//
//  AudioTester.swift
//  HoundSDK Swift Test Application
//
//  Created by Ken Huang on 11/25/16.
//  Copyright © 2016 SoundHound. All rights reserved.
//

import Foundation
import AVFoundation
import AudioUnit

enum AudioTesterError:Int {
	case none
	case permissionDenied
}

typealias AudioTesterDataHandler = (NSError?, Data?)->Void
typealias AudioTesterErrorHandler = (NSError?)->Void
typealias AudioTesterPermissionCallback = (Bool)->Void

let AudioTesterErrorDomain = "AudioTesterErrorDomain"

let INPUT_BUS:UInt32 = 1
let OUTPUT_BUS:UInt32 = 0

class AudioTester {

	static let instance:AudioTester = AudioTester()
	
	fileprivate var _audioUnit: AudioUnit? = nil
	fileprivate var _queue: DispatchQueue
	fileprivate var _session: AVAudioSession {
		get {
			return AVAudioSession.sharedInstance()
		}
	}
	
	fileprivate var _handler: AudioTesterDataHandler?
	
	fileprivate var streamDescription: AudioStreamBasicDescription {
		get {
			var audioStreamDescription: AudioStreamBasicDescription = AudioStreamBasicDescription()
			audioStreamDescription.mFormatID = kAudioFormatLinearPCM
			audioStreamDescription.mFormatFlags = kAudioFormatFlagIsSignedInteger | kAudioFormatFlagIsPacked
			audioStreamDescription.mFramesPerPacket = 1
			audioStreamDescription.mChannelsPerFrame = 1
			audioStreamDescription.mBitsPerChannel = 16
			audioStreamDescription.mBytesPerPacket = 2
			audioStreamDescription.mBytesPerFrame = 2
			audioStreamDescription.mSampleRate = self._session.sampleRate
			
			return audioStreamDescription
		}
	}
	
	init() {
		self._queue = DispatchQueue(label: "com.hound.audio", attributes: [])
	}
	
	func startAudioWithSampleRate(_ sampleRate:Double, handler: AudioTesterDataHandler?) {
		self.stopAudioWithHandler{ error in
			guard error == nil else {
				handler?(error, nil)
				return
			}
			
			self._queue.async {
				self._handler = handler
				self.requestPermissions { granted in
					
					guard granted else {
						handler?(self.error(.permissionDenied), nil)
						return
					}
					do {
						try self._session.setCategory(AVAudioSessionCategoryPlayAndRecord, with: .defaultToSpeaker)
						try self._session.setPreferredSampleRate(sampleRate)
						try self._session.setActive(true)
						
						var status: OSStatus = kAudioServicesNoError
						
						var audioComponentDescription: AudioComponentDescription = AudioComponentDescription()
						audioComponentDescription.componentType = kAudioUnitType_Output
						audioComponentDescription.componentSubType = kAudioUnitSubType_RemoteIO
						audioComponentDescription.componentManufacturer = kAudioUnitManufacturer_Apple
						audioComponentDescription.componentFlags = 0
						audioComponentDescription.componentFlagsMask = 0
						let audioComponent: AudioComponent? = AudioComponentFindNext(nil, &audioComponentDescription)
						
						guard audioComponent != nil else {
							handler?(self.error(kAudioServicesUnsupportedPropertyError),nil)
							return
						}
						
						status = AudioComponentInstanceNew(audioComponent!, &self._audioUnit)
						
						if status == kAudioServicesNoError {
							var yes: UInt32 = 1
							
							status = AudioUnitSetProperty(self._audioUnit!, kAudioOutputUnitProperty_EnableIO, kAudioUnitScope_Input, INPUT_BUS, &yes, UInt32(MemoryLayout.size(ofValue: yes)))
						}
						
						var streamDescription: AudioStreamBasicDescription = self.streamDescription
						
						if (status == kAudioServicesNoError) {
							status = AudioUnitSetProperty(self._audioUnit!, kAudioUnitProperty_StreamFormat, kAudioUnitScope_Output, INPUT_BUS, &streamDescription, UInt32(MemoryLayout.size(ofValue: streamDescription)))
						}
						
						if (status == kAudioServicesNoError) {
							status = AudioUnitSetProperty(self._audioUnit!, kAudioUnitProperty_StreamFormat, kAudioUnitScope_Input, OUTPUT_BUS, &streamDescription, UInt32(MemoryLayout.size(ofValue: streamDescription)))
						}
						
						if (status == kAudioServicesNoError) {
							var renderCallbackStruct = AURenderCallbackStruct(inputProc: audioTesterRenderCallback, inputProcRefCon: UnsafeMutableRawPointer(Unmanaged.passUnretained(self).toOpaque()))
							status = AudioUnitSetProperty(self._audioUnit!, kAudioUnitProperty_SetRenderCallback, kAudioUnitScope_Global, OUTPUT_BUS, &renderCallbackStruct, UInt32(MemoryLayout.size(ofValue: renderCallbackStruct)))
						}
						
						if (status == kAudioServicesNoError) {
							status = AudioUnitInitialize(self._audioUnit!)
						}
						
						if (status == kAudioServicesNoError) {
							status = AudioOutputUnitStart(self._audioUnit!)
						}
						
						if let error = self.error(status) {
							handler?(error, nil)
						}
						
					} catch let error {
						handler?(error as NSError, nil)
					}
				}
			}
		}
	}
	
	func stopAudioWithHandler(_ handler: AudioTesterErrorHandler?) {
		_queue.async {
			var error: NSError?
			guard self._audioUnit != nil else {
				handler?(error)
				return
			}
			
			var status:OSStatus
			status = AudioOutputUnitStop(self._audioUnit!)
			error = self.error(status)
			
			status = AudioUnitUninitialize(self._audioUnit!)
			
			error = error ?? self.error(status)
			
			status = AudioComponentInstanceDispose(self._audioUnit!)
			
			error = error ?? self.error(status)
			
			self._audioUnit = nil
			
			handler?(error)
		}
	}
	
	func render(_ actionFlags: UnsafeMutablePointer<AudioUnitRenderActionFlags>,
		timeStamp: UnsafePointer<AudioTimeStamp>,
		busNumber:UInt32,
		frameCount:UInt32,
		bufferList: UnsafeMutablePointer<AudioBufferList>?) -> OSStatus {
		
		var status = kAudioServicesNoError
		if bufferList == nil {
			status = kAudioServicesUnsupportedPropertyError
		} else {
			status = AudioUnitRender(self._audioUnit!, actionFlags, timeStamp, INPUT_BUS, frameCount, bufferList!)
			
			if status == kAudioServicesNoError {
				
				if (bufferList?.pointee.mNumberBuffers)! > 0 {
					let buffers = UnsafeMutableAudioBufferListPointer(bufferList)
					let buffer = buffers?.first!.mData
					let length = MemoryLayout<Int16>.size * Int(frameCount)
					
					let data = Data(bytes: buffer!, count: length)
					
					self._handler?(nil, data)
				}
			}
		}
		
		actionFlags.pointee = [actionFlags.pointee, AudioUnitRenderActionFlags.unitRenderAction_OutputIsSilence]
		
		let error = self.error(status)
		
		if error != nil {
			self._handler?(error, nil)
		}
		
		return status
	}
	
	fileprivate func error(_ status:OSStatus)->NSError? {
		var error: NSError?
		
		if (status != kAudioServicesNoError) {
			error = NSError(domain: NSOSStatusErrorDomain, code: Int(status), userInfo: nil)
		}
		
		return error
	}
	
	fileprivate func error(_ code: AudioTesterError) -> NSError? {
		return NSError(domain: AudioTesterErrorDomain, code: code.rawValue, userInfo: nil)
	}
	
	fileprivate func requestPermissions(_ completionHandler: @escaping AudioTesterPermissionCallback) {
		var permission: AVAudioSessionRecordPermission = .undetermined
		permission = self._session.recordPermission()
		
		switch (permission) {
        case AVAudioSessionRecordPermission.denied:
    		completionHandler(false)
    	case AVAudioSessionRecordPermission.granted:
    		completionHandler(true)
        default:
            self._session.requestRecordPermission(completionHandler)
    	}
	}
}

func audioTesterRenderCallback(_ context:UnsafeMutableRawPointer, actionFlags:UnsafeMutablePointer<AudioUnitRenderActionFlags>, timeStamp:UnsafePointer<AudioTimeStamp>, busNumber:UInt32, frameCount:UInt32, bufferList: UnsafeMutablePointer<AudioBufferList>?) -> OSStatus
{
	let audioTester = unsafeBitCast(context, to: AudioTester.self)
	return audioTester.render(actionFlags, timeStamp: timeStamp, busNumber: busNumber, frameCount: frameCount, bufferList: bufferList)
}
